File: src/murmur/pidfile.py
"""Pidfile mutual exclusion: one event tap / mic at a time across engines.

The native Swift engine (Murmur.app) and the Python engine must never both hold
a CGEventTap or the microphone. A single file ``~/.murmur/engine.pid`` records
the current owner; whoever is about to arm an event tap acquires it first and
refuses if a live *other* engine already holds it.

The file format is the cross-engine contract (mirrored in
``Murmur/Sources/Engine/Pidfile.swift``): line 1 is the pid, line 2 the owner
name (``python`` | ``native``). Pure stdlib so it unit-tests on any platform.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

PIDFILE = Path.home() / ".murmur" / "engine.pid"


@dataclass(frozen=True)
class Owner:
    pid: int
    name: str  # "python" | "native"


def _pid_alive(pid: int) -> bool:
    """True if a process with *pid* exists (POSIX ``kill(pid, 0)``)."""
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return False
    except PermissionError:
        return True  # exists, owned by another user
    return True


def read(path: Path = PIDFILE) -> Owner | None:
    """Parse the pidfile, or None if missing/empty/malformed."""
    try:
        text = path.read_text()
    except OSError:
        return None
    lines = text.splitlines()
    if not lines:
        return None
    try:
        pid = int(lines[0].strip())
    except ValueError:
        return None
    name = lines[1].strip() if len(lines) > 1 else ""
    return Owner(pid, name)


def current_holder(path: Path = PIDFILE) -> Owner | None:
    """The *live* owner of the pidfile, or None (no file, or a stale dead pid)."""
    owner = read(path)
    if owner is None or not _pid_alive(owner.pid):
        return None
    return owner


def acquire(name: str, pid: int | None = None, path: Path = PIDFILE) -> Owner | None:
    """Claim the pidfile for *name*.

    Returns None on success, or the live foreign holder that blocked the claim
    (the caller then refuses to arm). A file held by a live *other* process
    blocks; our own pid or a stale (dead) pid is overwritten. Written atomically
    (temp + rename).
    """
    me = os.getpid() if pid is None else pid
    holder = current_holder(path)
    if holder is not None and holder.pid != me:
        return holder
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.parent / f"{path.name}.{me}.tmp"
    tmp.write_text(f"{me}\n{name}\n")
    os.replace(tmp, path)
    return None


def release(name: str | None = None, pid: int | None = None, path: Path = PIDFILE) -> None:
    """Remove the pidfile if this pid still owns it (no-op otherwise)."""
    me = os.getpid() if pid is None else pid
    owner = read(path)
    if owner is not None and owner.pid == me:
        try:
            path.unlink()
        except OSError:
            pass

File: Pomvox/Sources/Engine/Pidfile.swift
import Foundation

/// Pidfile mutual exclusion — one event tap / mic at a time across the native
/// and Python engines. Mirror of `src/pomvox/pidfile.py`; the file format is the
/// cross-engine contract: line 1 = pid, line 2 = owner name ("native" |
/// "python"). The native engine acquires before arming and refuses if a live
/// Python engine already holds it.
struct Pidfile {
    struct Owner: Equatable {
        let pid: Int32
        let name: String
    }

    let url: URL

    static let defaultURL = URL(fileURLWithPath:
        NSString(string: "~/.pomvox/engine.pid").expandingTildeInPath)

    init(url: URL = Pidfile.defaultURL) { self.url = url }

    static func pidAlive(_ pid: Int32) -> Bool {
        if pid <= 0 { return false }
        if kill(pid, 0) == 0 { return true }
        return errno == EPERM  // exists, owned by another user
    }

    func read() -> Owner? {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
        guard let first = lines.first,
              let pid = Int32(first.trimmingCharacters(in: .whitespaces)) else { return nil }
        let name = lines.count > 1 ? String(lines[1]).trimmingCharacters(in: .whitespaces) : ""
        return Owner(pid: pid, name: name)
    }

    /// The live owner, or nil (no file / stale dead pid).
    func currentHolder() -> Owner? {
        guard let owner = read(), Self.pidAlive(owner.pid) else { return nil }
        return owner
    }

    /// Claim for `name`. Returns nil on success, or the live foreign holder that
    /// blocked the claim (the caller then refuses to arm). A live *other* pid
    /// blocks; our own pid or a stale dead pid is overwritten. Atomic write.
    @discardableResult
    func acquire(_ name: String,
                 pid: Int32 = ProcessInfo.processInfo.processIdentifier) -> Owner? {
        if let holder = currentHolder(), holder.pid != pid { return holder }
        try? FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? "\(pid)\n\(name)\n".write(to: url, atomically: true, encoding: .utf8)
        return nil
    }

    /// Remove the pidfile if this pid still owns it (no-op otherwise).
    func release(pid: Int32 = ProcessInfo.processInfo.processIdentifier) {
        guard let owner = read(), owner.pid == pid else { return }
        try? FileManager.default.removeItem(at: url)
    }
}
